tasking/pulp/tasking/base.py

```diff
@@ -0,0 +1,444 @@
+    and keyword args for the inner task. These arguments are named inner_args and inner_kwargs.
+    inner_args is a list, and inner_kwargs is a dictionary passed to the inner task as positional
+    and keyword arguments using the * and ** operators.
+
+    The inner task is dispatched into a dedicated queue for a worker that is decided at dispatch
+    time. The logic deciding which queue receives a task is controlled through the
+    find_worker function.
+
+    :param name:          The name of the task to be called
+    :type name:           basestring
+    :param inner_task_id: The UUID to be set on the task being called. By providing
+                          the UUID, the caller can have an asynchronous reference to the inner task
+                          that will be dispatched.
+    :type inner_task_id:  basestring
+    :param resource_id:   The name of the resource you wish to reserve for your task. The system
+                          will ensure that no other tasks that want that same reservation will run
+                          concurrently with yours.
+    :type  resource_id:   basestring
+
+    :return: None
+    """
+    while True:
+        # Find a worker who already has this reservation, it is safe to send this work to them
+        try:
+            worker = ReservedResource.objects.get(resource=resource_id).worker
+        except ReservedResource.DoesNotExist:
+            pass
+        else:
+            break
+
+        try:
+            worker = _get_unreserved_worker()
+        except Worker.DoesNotExist:
+            pass
+        else:
+            break
+
+        # No worker is ready for this work, so we need to wait
+        time.sleep(0.25)
+
+    task_status = TaskStatus.objects.get(pk=task_id)
+    ReservedResource.objects.create(task=task_status, worker=worker, resource=resource_id)
+
+    inner_kwargs['routing_key'] = worker.name
+    inner_kwargs['exchange'] = DEDICATED_QUEUE_EXCHANGE
+    inner_kwargs['task_id'] = task_id
+
+    try:
+        celery.tasks[name].apply_async(*inner_args, **inner_kwargs)
+    finally:
+        _release_resource.apply_async((task_id, ), routing_key=worker.name,
+                                      exchange=DEDICATED_QUEUE_EXCHANGE)
+
+
+def _get_unreserved_worker():
+    """
+    Find an unreserved Worker
+
+    Return the Worker instance that has no ReservedResource associated with it. If all workers have
+    ReservedResource relationships a Worker.DoesNotExist pulp.app.model.Worker exception is
+    raised.
+
+    This function also provides randomization for worker selection.
+
+    :raises Worker.DoesNotExist: If all workers have ReservedResource entries associated with them.
+
+    :returns:          The Worker instance that has no reserved_resource
+                       entries associated with it.
+    :rtype:            pulp.app.model.Worker
+    """
+    free_workers = [Worker.objects.annotate(Count('reservations')).filter(reservations__count=0)]
```

Comment:
I haven't used ``annotate`` before, but if I'm understanding correctly, ``filter(...)`` should return a query set, right? And you want to evaluate that into a list? If so, I think the ``[...]`` notation will not do what you want, and will instead give you a list whose only member is a query set. I think you want ``list(...)`` here.